docs/architecture/adr-002-ibc-relayer.md

```diff
@@ -0,0 +1,644 @@
+
+```$xslt
+[[connections]]
+
+[connections.src]
+  client_id = "clA1"
+  connection_id = "conn1-idA-clA1"
+[connections.dest]
+  client_id = "clB1"
+  connection_id = "conn1-idB-clB1"
+
+[[connections.paths]]
+  src_port = "app1-port-A"
+  dest_port = "app1-port-B"
+  direction = "unidirectional"
+  [[connections.paths]]
+  src_port = "app2-port-A"
+  dest_port = "app2-port-B"
+  direction = "bidirectional"
+```
+
+### Relayer Commands
+
+`relayer start` starts the relayer with the specified configuration file
+
+`relayer verify config` reads and verifies that the specified configuration file parses correctly
+
+Other commands will be added as required.
+
+_Note: it is expected that the chains can be configured with client, connections and channels. Integration testing will be done with Gaia application, therefore access to the gaiacli for IBC related configuration is available_
+
+### Relayer Queries
+The relayer queries chain state in order to build the IBC messages. It is expected that each chain type provides implementations of these queries. Initial Rust relayer implementation will be tested with Cosmos-SDK/Tendermint chains, and while some of the IBC-Modules functionality in Rust is not required (e.g. handler functions), a "query" crate should be available for the relayer. 
+For tendermint, the queries use the `abci.RequestQuery` over rpc/http to retrieve the data. 
+
+The format of the public/ provable state query parameters and responses is chain independent and should also be defined in this crate.
+ 
+The following queries are required:
+
+- `query_store_prefix(chain)` - returns the commitment prefix of the chain (returns chain specific []byte, e.g. `ibc` for tendermint)
+- `query_all_client_states(chain)` - returns the IBC light clients instantiated on the chain
+- `query_client_consensus_state(chain, clientID, height)` - returns the consensus state proof for a light client at a given height if height > 0, else it returns the latest height
+- `query_connections(chain)` - returns all connections created on the chain
+- `query_client_connections(chain, clientID)` - returns all connections associated with a light client
+- ...more to be added
+
+
+### Relayer Algorithm
+A relayer algorithm is described in [relayer algorithm described in IBC Specifigication](https://github.com/cosmos/ics/blame/master/spec/ics-018-relayer-algorithms/README.md#L47) and [Go relayer implementation ](https://github.com/cosmos/relayer/blob/f3a302df9e6e0c28883f5480199d3190821bcc06/relayer/strategies.go#L49.).
+
+This section describes the details of the relayer algorithm in the Rust implementation. 
+The purpose of the relayer algorithm is to build and forward messages over the `src->dest` relaying paths it has been configured for. It is executed periodically with the frequency specified in the configuration file, with each run being a **relay cycle**.
+For every destination the relayer:
```

Comment:
Not clear what is meant by destination here. Is it for every chain (defined in config file)?